### PR TITLE
feat(package): .serialize() also writes package-lock version

### PR DESCRIPTION
**Pull request by mikemimik**
## Description

The `@core/package` class has a method called `.serialize()` which writes the in-memory package to disk. When running `lerna version` the `.version` property of the package is updated, written, and committed (depending on flags). However, the `package-lock.json` file _(which also has a top level `version` property)_ doesn't get updated, and is then out of sync with the `package.json` file.

This change updates the `serialize()` function to check if the package has a `package-lock.json` file in the directory, and if it does: read it in, update the version, then write it to disk.

## Motivation and Context

Any monorepo project that has `package-lock.json` files in it's package folders need to run another step to get the `package-lock.json` files to stay in sync with the `package.json` files. The script usually looks something like:

```bash
lerna exec -- npm install --no-audit --no-progress
git add packages/**/package-lock.json
git commit -m "chore(package-lock): bump package lock files [skip ci]"
git push
```
_(or some such variation of the above)_

Solves: https://github.com/lerna/lerna/issues/1998
Linked: https://github.com/lerna/lerna/issues/1415

## How Has This Been Tested?

Tests where added to the `core/package/__tests__/core-package.test.js` file.
- `describe("get .lockFileLocation")` was **added**
  - contains `it("should return the package-lock location")`
- `describe(".serialize()")` was **updated**
  - contains `it("writes changes to package.json to disk")` **updated**
  - contains `it("writes changes to package-lock to disk")` **added**

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by mikemimik** [comment]
It occurs to me that while this update will bump the version property in the `package-lock.json` file to be the same as the version property in the `package.json` file, the `package-lock.json` tree structure does not reflect the changes in package versions. This is a fundamental flaw in this feature work, and thusly should not be merged in yet. I will work to remedy this.

File: core/package/index.js [hunk 1]
```diff
@@ -77,6 +86,9 @@ class Package {
       manifestLocation: {
         value: path.join(location, "package.json"),
       },
+      lockFileLocation: {
+        value: path.join(location, "package-lock.json"),
```
Comment (evocateur) [inline core/package/index.js]: this could be `npm-shrinkwrap.json`, or `yarn.lock`, or, or, or...

File: core/package/index.js [hunk 2]
```diff
@@ -181,7 +193,27 @@ class Package {
...
+    const lockChanges = { version: this[PKG].version };
+
+    const tasks = [
+      () => loadJsonFile(this.lockFileLocation),
+      lockFileData => Promise.resolve(Object.assign({}, lockFileData, lockChanges)),
+      updatedLockFileData =>
+        writeJsonFile(this.lockFileLocation, updatedLockFileData, { detectIndent: true }),
+    ];
+
+    return hasPackageLock(this.lockFileLocation) ? pWaterfall(tasks) : Promise.resolve();
```
Comment (evocateur) [inline core/package/index.js]: I don't think we need `p-waterfall` just to chain promises. This check should be done way earlier, too.

File: package.json [hunk 3]
```diff
@@ -7,7 +7,7 @@
     "ci": "npm test -- --ci --maxWorkers=2 && npm run integration -- --ci",
     "fix": "npm run lint -- --fix",
     "integration": "jest --config jest.integration.js --maxWorkers=2",
-    "lint": "eslint . --ignore-path .gitignore --cache --cache-location ./node_modules/.cache/eslint",
+    "lint": "eslint . --ignore-path .gitignore --ignore-pattern \"**/node_modules/\" --cache --cache-location ./node_modules/.cache/eslint",
```
Comment (evocateur) [inline package.json]: Unclear why this is necessary, as it is already ignored.

File: core/package/index.js [hunk 4]
```diff
@@ -30,6 +33,12 @@ function shallowCopy(json) {
   }, {});
 }
 
+// determine if a package-lock.json file exists
```
Comment (evocateur) [inline core/package/index.js]: This isn't the appropriate place for lockfile maintenance. It's a function of the version command, not integral to the nature of a Package.

**Comment by evocateur** [comment]
[v3.18.2](https://github.com/lerna/lerna/releases/tag/v3.18.2)

**Pull request closed without merging**